index.html

```diff
@@ -101,19 +207,655 @@ <h3>
+
+      <p>
+      Instances of {{UDPSocket}} are created with the internal slots described in the following table:
+
+      <table class="simple" data-dfn-for="UDPSocket">
+        <tr>
+          <th>Internal slot
+          <th>Initial value
+          <th>Description (non-normative)
+        <tr>
+          <td><dfn>[[\readable]]</dfn>
+          <td>`null`
+          <td>A {{ReadableStream}} that receives data from the socket 
+        <tr>
+          <td><dfn>[[\writable]]</dfn>
+          <td>`null`
+          <td>A {{WritableStream}} that transmits data to the socket
+        <tr>
+          <td><dfn>[[\openedPromise]]</dfn>
+          <td>`new Promise`
+          <td>A {{Promise}} used to wait for the socket to be opened. Corresponds to the {{UDPSocket/opened}} member.
+        <tr>
+          <td><dfn>[[\closedPromise]]
+          <td>`new Promise`
+          <td>A {{Promise}} used to wait for the socket to close or error. Corresponds to the {{UDPSocket/closed}} member.
+      </table>
+
+      <section>
+        <h3><dfn>constructor()</dfn> method</h3>
+
+        <aside class="example">
+          In order to communicate via UDP a socket must be opened first. The socket object constructor
+          allowes the site to specify the necessary parameters which control how data is transmitted and received.
+
+          <pre class="js">
+            const socket = new UDPSocket(/*remoteAddress=*/, /*remotePort=*/, /*options=*/);
+          </pre>
+
+          The developer should wait for the {{UDPSocket/opened}} promise to be resolved to gain access to {{UDPSocketOpenInfo/readable}} and {{UDPSocketOpenInfo/writable}} streams.
+
+          <pre class="js">
+            const { readable, writable } = await socket.opened;
+          </pre>
+
+          Once {{UDPSocket/opened}} has resolved, the {{UDPSocketOpenInfo/readable}} and {{UDPSocketOpenInfo/writable}}
+          can be accessed to get the {{ReadableStream}} and {{WritableStream}} instances for receiving data from and sending data to the socket.
+        </aside>
+
+
+        The {{UDPSocket/constructor()}} steps are:
+
+        <ol>
+          <li>If [=this=]'s [=relevant global object=]'s [=associated Document=] is
+        not [=allowed to use=] the [=policy-controlled feature=] named "[=policy-controlled feature/direct-sockets=]", throw a
+        "{{NotAllowedError}}" {{DOMException}}.
+          <li>If |options|["{{UDPSocketOptions/sendBufferSize}}"] is equal to 0, throw a {{TypeError}}.
+          <li>If |options|["{{UDPSocketOptions/receiveBufferSize}}"] is equal to 0, throw a {{TypeError}}.
+          <li>If |options|["{{UDPSocketOptions/readableStreamBufferSize}}"] is equal to 0, throw a {{TypeError}}.
+          
+          <li>Perform the following steps [=in parallel=].
+            <ol>
+              <li>
+                Invoke the operating system to open the UDP socket using the given |remoteAddress| and |remotePort| and the 
```

Comment:
We should probably be more specific about how the socket is bound and connected, since those are important for UDP. Right now it looks like the constructor parameters only cover how we call connect(2). How are the local address and port selected?

Reply:
In Chromium `Connect(...)` [doesn't allow](https://source.chromium.org/chromium/chromium/src/+/main:services/network/public/mojom/udp_socket.mojom;drc=75425ebb5303c7bb416b2261e9b27114ec903d3f;l=103) us to specify `localAddress`/`localPort`. I guess we can say that these are also implementation-defined and chosen automatically?

Reply:
It looks like `Bind()` and `Connect()` are mutually exclusive in the Chromium network library but it doesn't seem like that is case for the underlying operating system APIs. You could call bind() to select a local port and then call connect() to set the default destination address (as well as filter incoming packets).

The current API seems to preclude creating an unconnected UDP socket. We should probably fix that, or at least make it possible to add support for that in a backwards-compatible way. I think what probably makes sense is to move the `remoteAddress` and `remotePort` parameters into the `UDPSocketOptions` dictionary and add `localAddress` and `localPort` options as well. Then the constructor steps can decide whether to call `Bind()` or `Connect()` depending on which parameters are provided. We can, for the time being, not support specifying both local and remote endpoint information if the Chromium network library doesn't support it.

It looks like the way that the `chrome.sockets.udp` API handled this was that it didn't support calling connect() on a UDP socket so they were always connectionless and required specifying the remote address and port with every packet.

Reply:
While I like the idea with moving both pairs of params (`remote...`/`local...`) to the options dictionary, adding support for `Bind(...)` comes with some questions:

- What about the security concerns? Technically `SendTo(...)` will follow the same logic with respect to address resolving, so it might not be a real issue (but what about `Bind(...)` -- are there any restrictions on what we can/cannot bind to?) 
- The way `chrome.sockets.udp` sends packets (via `SendTo(...)`) is that it first attempts to resolve the given address without any caching (that's horrible in terms of performance given the number of overhead mojo calls). Do we need to think about this?
- What about sending errors? Since we're planning to send data to various independent remotes, one error should not be able to close the `WritableStream`.

Reply:
> * What about the security concerns? Technically `SendTo(...)` will follow the same logic with respect to address resolving, so it might not be a real issue (but what about `Bind(...)` -- are there any restrictions on what we can/cannot bind to?)

For the moment I think we can consider that out of scope. The overall stance of this API as currently specified is that it is up to the user agent to provide restrictions (if any) on where packets can be sent and I think that should also apply to how sockets can be bound.

> * The way `chrome.sockets.udp` sends packets (via `SendTo(...)`) is that it first attempts to resolve the given address without any caching (that's horrible in terms of performance given the number of overhead mojo calls). Do we need to think about this?

The way that the `chrome.sockets.udp` API handles this is terrible from a performance perspective. I think a `remoteAddress` specified on an individual packet should be restricted to an IP address, which matches the behavior of the underlying sendto(2) function. 

> * What about sending errors? Since we're planning to send data to various independent remotes, one error should not be able to close the `WritableStream`.

I'm not sure (aside from DNS resolution, which I think we should disallow) that calling `SendTo(...)` introduces any new kinds of errors we couldn't already encounter with a regular `Send(...)`. That said, this does expose a weakness in trying to model sending UDP packets as a stream because each send is logically distinct, compared to a TCP socket.

In the interest of getting this PR merged let's make what seems like the minimal change necessary to make support an unconnected socket backwards compatible and move the `remoteAddress` and `remotePort` parameters into `UDPSocketOptions`. They can, for now, be marked as required parameters.

Reply:
Well, raises some concerns to think about :)
I moved the two options into the dictionary -- please look.